### PR TITLE
Update pyproject and workflow files

### LINKED ISSUE #267: Use `pyproject.toml` and `requirements.txt` files with `pip-tools`
At the moment, it uses `setup.cfg` with `requirements.in`, which is pretty old style.

### LINKED ISSUE #269: Update workflow files to use our current standard of when to run them
See https://github.com/ImperialCollegeLondon/QwikApp_web as an example to follow.

### PR DESCRIPTION
**Pull request by dalonsoa**
Mainly, the dependencies have been included in the pyproject.toml file and the approrpiate requirements file re-created with pip-compile. 

Also, the workflow files for GitHub actions have been updated to follow our current approach and dependabot file has been added. Also, pre-commit.ci has been enabled in this repo.

Close #267 
Close #269

### CODE REVIEW

**Review by AdrianDAlessandro — COMMENTED** [review]
I'm getting the same bootstrap errors. But I also have some more general comments about the changes to the pyproject.toml and workflows.

**Review by Sahil590 — APPROVED** [review]
LGTM. I do get some bootstrap errors but can see it will be fixed in a different PR.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -29,64 +28,18 @@ jobs:
...
-      uses: docker/build-push-action@v2
-      with:
-        context: .
-        push: true
-        tags: ${{ steps.meta.outputs.tags }}
-
-  docs:
-    needs: test
-    timeout-minutes: 15
-    runs-on: ubuntu-latest
```
Comment (AdrianDAlessandro) [inline .github/workflows/ci.yml:78]: It's probably useful to have a test-docs step here that runs `mkdocs build` just to make sure the docs build successfully so they don't fail on release.

**Review by AdrianDAlessandro — APPROVED** [review]
This seems all good now. I've made a suggestion to run an docs build test in the ci.yml to confirm they will build successfully.

Also, you should update the required tests in the branch protections after merging this.

Also, also, what are those CodeQL workflows that are running? I can't work out how they're configured.

**Comment by dalonsoa** [comment]
> Also, also, what are those CodeQL workflows that are running? I can't work out how they're configured.

No idea. GitHub started to run them at some point without me doing anything.

**Pull request merged**